keras/models.py

```diff
@@ -81,6 +81,8 @@ def weighted(y_true, y_pred, weights, mask=None):
         # score_array has ndim >= 2
         score_array = fn(y_true, y_pred)
         if mask is not None:
+            # Cast the mask to floatX to avoid float64 upcasting in theano
+            mask = K.cast(mask, K.floatx())
```

Comment:
float32 + int32 will upcast following c casting rules. This is normal. not
all int32 fit exactly in a float32. So not doing the upcast have problems
in some corner case.

Theano casting rules is very close to c casting rules. int8 and int16 won't
cause this upcast.

The current GPU back-end only support float32, so this can cause is some
corner case extra transfer.

Making the mask int16 would be better then int32, as the upcast could
happen elsewhere where that int32 mask is used.

On Thu, Jan 7, 2016 at 3:24 PM, Ozan Çağlayan <notifications@github.com>
wrote:

> In keras/models.py
> <https://github.com/fchollet/keras/pull/1419#discussion_r49121301>:
>
> > @@ -81,6 +81,8 @@ def weighted(y_true, y_pred, weights, mask=None):
> >          # score_array has ndim >= 2
> >          score_array = fn(y_true, y_pred)
> >          if mask is not None:
> > +            # Cast the mask to floatX to avoid float64 upcasting in theano
> > +            mask = K.cast(mask, K.floatx())
>
> Well I'm not an expert of keras nor theano. I also sent an email to
> theano-users list today but didn't receive anything yet: On my system a
> multiplication of float32 and int32 in theano results in creation of a
> float64 tensor. Here the mask is int32 and the score_array is float32.
>
> Are casts causing transfer between host and device?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/pull/1419/files#r49121301>.
>